kubernetes-addons/memory-bandwidth-exporter/collector/collector.go

```diff
@@ -0,0 +1,177 @@
+package collector
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Namespace defines the common namespace to be used by all metrics.
+const namespace = "rdt"
+
+var (
+	scrapeSuccessDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "scrape", "collector_success"),
+		"memory_bandwidth_exporter: Whether a collector succeeded.",
+		[]string{"collector"},
+		nil,
+	)
+)
+
+const (
+	defaultEnabled  = true
+	defaultDisabled = false
+)
+
+var (
+	factories              = make(map[string]func(logger log.Logger, interval time.Duration) (Collector, error))
+	initiatedCollectorsMtx = sync.Mutex{}
+	initiatedCollectors    = make(map[string]Collector)
+	collectorState         = make(map[string]*bool)
+	collectorMetrics       = make(map[string]string)
+)
+
+func registerCollector(collector string, isDefaultEnabled bool, factory func(
+	logger log.Logger, interval time.Duration) (Collector, error)) {
+	collectorState[collector] = &isDefaultEnabled
+	factories[collector] = factory
+}
+
+func ParseCollectorMetrics() bool {
+	isNeedNRIPlugin := false
+	for collector := range collectorState {
+		if collector == containerCollectorSubsystem {
+			var isDefaultEnabled bool
+			if *containerCollectorMetrics == noMetrics {
+				isDefaultEnabled = false
+			} else {
+				isDefaultEnabled = true
+			}
+			collectorState[collector] = &isDefaultEnabled
+			collectorMetrics[collector] = *containerCollectorMetrics
+			if *containerCollectorMetrics == allMetrics || strings.Contains(*containerCollectorMetrics, "mb") ||
+				strings.Contains(*containerCollectorMetrics, "llc") {
+				isNeedNRIPlugin = true
+			}
+		}
+		if collector == classCollectorSubsystem {
+			var isDefaultEnabled bool
+			if *classCollectorMetrics == "none" {
+				isDefaultEnabled = false
+			} else {
+				isDefaultEnabled = true
+			}
+			collectorState[collector] = &isDefaultEnabled
+			collectorMetrics[collector] = *classCollectorMetrics
+		}
+		if collector == nodeCollectorSubsystem {
+			var isDefaultEnabled bool
+			if *nodeCollectorMetrics == "none" {
+				isDefaultEnabled = false
+			} else {
+				isDefaultEnabled = true
+			}
+			collectorState[collector] = &isDefaultEnabled
+			collectorMetrics[collector] = *nodeCollectorMetrics
+		}
+	}
+	return isNeedNRIPlugin
+}
+
+// Collector implements the prometheus.Collector interface.
+type MBCollector struct {
+	Collectors map[string]Collector
+	logger     log.Logger
+}
+
+// NewCollector creates a new Collector.
+func NewCollector(logger log.Logger, interval time.Duration, filters ...string) (*MBCollector, error) {
+	f := make(map[string]bool)
+	for _, filter := range filters {
+		enabled, exist := collectorState[filter]
+		if !exist {
+			return nil, fmt.Errorf("missing collector: %s", filter)
+		}
+		if !*enabled {
+			return nil, fmt.Errorf("disabled collector: %s", filter)
```

Comment:
is this an error? A collector with no filters?

Reply:
Do you want to add a comment about performance overhead, like O(N) in the number of containers, some K for the filters (llc, memory local ..), some m for periodicity and amount of space needed on disk to hold metrics before it is pumped out.